### ISSUE TITLE
Missao 05

### ISSUE THREAD

**Issue by Adrielle-Araujo** [Adrielle-Araujo/MLCB#7]
**Exercício 1:**

<img width="1097" height="511" alt="Image" src="https://github.com/user-attachments/assets/eaae8638-bf32-4442-bcae-110ca15488cc" />

**Exercício 2:**

**CÓDIGO**

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score
import matplotlib.pyplot as plt
import seaborn as sns

data = {
    'Ano': [2018, 2015, 2019, 2020, 2017, 2014, 2018, 2016, 2021, 2013, 2019, 2017, 2015, 2020, 2018, 2016, 2022, 2014, 2019, 2017],
    'Quilometragem': [50000, 120000, 30000, 15000, 70000, 150000, 60000, 95000, 5000, 180000, 40000, 80000, 110000, 20000, 55000, 105000, 2000, 160000, 35000, 75000],
    'Potencia_Motor': [1.6, 2.0, 1.0, 1.4, 1.8, 2.0, 1.6, 1.8, 1.0, 2.2, 1.4, 1.6, 2.0, 1.0, 1.8, 2.0, 1.2, 2.2, 1.4, 1.8],
    'Tipo_Combustivel': ['Flex', 'Gasolina', 'Flex', 'Flex', 'Gasolina', 'Diesel', 'Flex', 'Gasolina', 'Flex', 'Diesel', 'Flex', 'Flex', 'Gasolina', 'Flex', 'Gasolina', 'Diesel', 'Flex', 'Diesel', 'Flex', 'Gasolina'],
    'Cambio': ['Manual', 'Automatico', 'Manual', 'Manual', 'Automatico', 'Manual', 'Automatico', 'Automatico', 'Manual', 'Automatico', 'Manual', 'Manual', 'Automatico', 'Manual', 'Automatico', 'Automatico', 'Manual', 'Automatico', 'Manual', 'Automatico'],
    'Preco': [45000, 38000, 55000, 65000, 52000, 35000, 48000, 46000, 75000, 32000, 58000, 50000, 40000, 68000, 53000, 44000, 82000, 30000, 60000, 51000]
}
df_carros = pd.DataFrame(data)

df_carros.loc[5, 'Quilometragem'] = np.nan
df_carros.loc[11, 'Potencia_Motor'] = np.nan

print("--- Análise Inicial: Dataset de Carros ---")
df_carros.info()
print("\nValores ausentes:")
print(df_carros.isnull().sum())

X = df_carros.drop('Preco', axis=1)
y = df_carros['Preco']

X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.2, random_state=42)

colunas_numericas = ['Ano', 'Quilometragem', 'Potencia_Motor']
colunas_categoricas = ['Tipo_Combustivel', 'Cambio']

pipeline_numerico = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

pipeline_categorico = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', pipeline_numerico, colunas_numericas),
        ('cat', pipeline_categorico, colunas_categoricas)
    ])

modelo_reg = LinearRegression()

pipeline_final_reg = Pipeline(steps=[('preprocessor', preprocessor),
                                     ('regressor', modelo_reg)])

pipeline_final_reg.fit(X_treino, y_treino)

print("\n Modelo de Regressão treinado com sucesso!")

previsoes_preco = pipeline_final_reg.predict(X_teste)

mae = mean_absolute_error(y_teste, previsoes_preco)
r2 = r2_score(y_teste, previsoes_preco)

print("\n--- Resultados da Avaliação (Regressão) ---")
print(f"Erro Absoluto Médio (MAE): R$ {mae:.2f}")
print(f"Score R²: {r2:.2f}")

plt.figure(figsize=(10, 6))
plt.scatter(y_teste, previsoes_preco)
plt.plot([y_teste.min(), y_teste.max()], [y_teste.min(), y_teste.max()], '--r', linewidth=2)
plt.xlabel("Preços Reais")
plt.ylabel("Preços Previstos")
plt.title("Preços Reais vs. Previsões do Modelo de Regressão")
plt.show()

<img width="944" height="413" alt="Image" src="https://github.com/user-attachments/assets/518ebccb-13ba-4475-9db8-98aca38af75c" />

<img width="1205" height="508" alt="Image" src="https://github.com/user-attachments/assets/d1956cca-5d24-4b0a-8905-78b7c1c33988" />

**Exercício 3:**

**CÓDIGO**

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import seaborn as sns
import matplotlib.pyplot as plt

data = {
    'Idade': [25, 45, 35, 50, 22, 55, 41, 28, 33, 60, 39, 48, 29, 52, 31, 65, 27, 38, 43, 58],
    'Plano_Mensal': [70, 95, 80, 105, 50, 110, 90, 65, 75, 115, 85, 100, 72, 108, 78, 120, 68, 82, 92, 112],
    'Meses_Contrato': [1, 24, 12, 36, 2, 48, 18, 5, 8, 60, 15, 30, 3, 40, 6, 55, 4, 10, 22, 50],
    'Tipo_Contrato': ['Mensal', 'Anual', 'Mensal', 'Dois anos', 'Mensal', 'Dois anos', 'Anual', 'Mensal', 'Mensal', 'Dois anos', 'Anual', 'Dois anos', 'Mensal', 'Dois anos', 'Mensal', 'Dois anos', 'Mensal', 'Anual', 'Anual', 'Dois anos'],
    'Suporte_Tecnico': ['Sim', 'Nao', 'Sim', 'Nao', 'Nao', 'Nao', 'Sim', 'Sim', 'Nao', 'Nao', 'Sim', 'Nao', 'Sim', 'Nao', 'Sim', 'Nao', 'Sim', 'Sim', 'Nao', 'Nao'],
    'Churn': ['Sim', 'Nao', 'Sim', 'Nao', 'Sim', 'Nao', 'Nao', 'Sim', 'Sim', 'Nao', 'Nao', 'Nao', 'Sim', 'Nao', 'Sim', 'Nao', 'Sim', 'Nao', 'Nao', 'Nao']
}
df_churn = pd.DataFrame(data)

df_churn.loc[3, 'Plano_Mensal'] = np.nan

print("--- Análise Inicial: Dataset de Churn ---")
df_churn.info()
print("\nValores ausentes:")
print(df_churn.isnull().sum())

X = df_churn.drop('Churn', axis=1)
y = df_churn['Churn']

X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.25, random_state=42)

colunas_numericas = ['Idade', 'Plano_Mensal', 'Meses_Contrato']
colunas_categoricas = ['Tipo_Contrato', 'Suporte_Tecnico']

pipeline_numerico = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])
pipeline_categorico = Pipeline(steps=[
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', pipeline_numerico, colunas_numericas),
        ('cat', pipeline_categorico, colunas_categoricas)
    ])

modelo_cla = LogisticRegression()

pipeline_final_cla = Pipeline(steps=[('preprocessor', preprocessor),
                                     ('classifier', modelo_cla)])

pipeline_final_cla.fit(X_treino, y_treino)

print("\n Modelo de Classificação treinado com sucesso!")

previsoes_churn = pipeline_final_cla.predict(X_teste)

acuracia = accuracy_score(y_teste, previsoes_churn)
relatorio = classification_report(y_teste, previsoes_churn)
matriz_conf = confusion_matrix(y_teste, previsoes_churn)

print("\n--- Resultados da Avaliação (Classificação) ---")
print(f"Acurácia: {acuracia*100:.2f}%")
print("\nRelatório de Classificação:")
print(relatorio)

plt.figure(figsize=(8, 6))
sns.heatmap(matriz_conf, annot=True, fmt='d', cmap='Blues', xticklabels=['Nao Churn', 'Churn'], yticklabels=['Nao Churn', 'Churn'])
plt.xlabel('Previsto')
plt.ylabel('Real')
plt.title('Matriz de Confusão')
plt.show()

<img width="1138" height="551" alt="Image" src="https://github.com/user-attachments/assets/0da6c7f6-7afd-443f-8776-7fe1d4e0467e" />
